README.md
Chef Cookbook - Linux Base
==============
This Chef cookbook will configure some basic system settings, add/remove some common packages, and deploy standard configurations for some common services  

Changes Performed
------------
1. Remove the broken Fedora Rawhide repo from CentOS 
2. Make sure CentOS has EPEL
3. Remove the following packages from Ubuntu-based systems
    1. landscape-client-ui
    2. landscape-client-ui-install
    3. landscape-client
    4. landscape-common
    5. update-motd
4. Install the following packages on Ubuntu-based systems
    1. bmon
    2. htop
    3. vim
    4. curl
    5. gnupg2
    6. atop
    7. git
    8. molly-guard
    9. python-pip
    10. python-dev
    11. python3-pip
    12. unzip
    13. psmisc
5. Install the following packages on CentOS-based systems
    1. vim-enhanced
    2. curl
    3. htop
    4. gnupg2
    5. git
    6. lsof
    7. telnet
    8. net-tools
    9. bmon
    10. atop
    11. wget
    12. python2-pip
    13. python34-pip
    14. python34-devel
    15. unzip
    16. gcc
    17. psmisc
    18. bind-utils
6. Make sure the `/tools` directory exists and is accessible only to root
7. install the following python2.7 packages
    1. urllib3
    2. pyasn1
8. Remove root's .bashrc on Ubuntu-based systems
9. Deploy a custom system-wide bashrc and a supporting git.bash file for working in git repos
10. Install thefuck in python3 and configure it with [Matthew Ahrenstein](https://www.ahrenstein.com)'s [custom rules](https://github.com/ahrenstein/thefuck-rules)
11. Install and configure ntp to run off the NTP pool using the America/New_York timezone

Requirements
------------
1. Chef (Tested on Chef 13.6.4)
2. Linux chef-clients (Tested on Ubuntu 14.04, Ubuntu 16.04, Ubuntu 18.04, and CentOS 7.2 but kitchen will let you test anything you want)
3. The apt cookbook from Chef Supermarket
4. The yum-epel cookbook from Chef Supermarket

Installation Tips
------------

1. We personally use Berks to install this into Chef servers, because it will grab the dependencies for you as well.

Limitations
------------
 1. None so far :)

Known Issues
------------
1. None so far :)

Use Cases
------------
Getting a minimum base configuration going on new Linux servers 

Donate To Support This Chef Cookbook
------------
Route 1337, LLC operates entirely on donations. If you find these scripts useful, please consider [contacting us](https://www.route1337.com/contact-us/) about how to donate.

Thank you for your support!


CHANGELOG.md
Linux Base - Changelog
==============
A list of all the changes made to this cookbook

Version 0.1.7
------------

1. Removing stalecryptocurrency wallets from donation section of README

Version 0.1.6
------------

1. Added testing to verify Ubuntu 18.04 support

Version 0.1.5
------------

1. metadata.rb has been changed to set 13.6.4 as the minimum chef-client version instead of the only version

Version 0.1.4
------------

1. Fixed python repo issues with thefuck for Ubuntu 14.04

Version 0.1.3
------------

1. We now remove a broken Fedora Rawhide from CentOS machines
2. Kitchen is now locked to testing on the version of Chef we use in production

Version 0.1.2
------------

1. Bringing documentation in line with Chef Supermarket Foodcritic expectations

Version 0.1.1
------------

1. Added psmisc package to all supported operating systems
2. Added bind-utils to CentOS

Version 0.1.0
------------

1. Initial Release


spec/unit/recipes/ntp_spec.rb
#
# Cookbook Name:: r1337-linux-base
# Spec:: ntp
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

require 'spec_helper'

describe 'r1337-linux-base::ntp' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/bash_spec.rb
#
# Cookbook Name:: r1337-linux-base
# Spec:: bash
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

require 'spec_helper'

describe 'r1337-linux-base::bash' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/smoke/default/ntp_test.rb
#
# Cookbook Name:: r1337-linux-base
# Spec:: ntp
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

# # encoding: utf-8

# Inspec test for recipe r1337-linux-base::ntp

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

if os[:name] == 'ubuntu'

  # Verify the NTP config exists
  describe file('/etc/ntp.conf') do
    it { should exist }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0644 }
  end

  # Verify the ntp package is installed
  describe package('ntp') do
    it { should be_installed }
  end

  # Verify the localtime is set correctly
  describe file('/etc/localtime') do
    it { should be_symlink }
    it { should be_linked_to '/usr/share/zoneinfo/posixrules' } # This is the final link for /usr/share/zoneinfo/America/New_York on Ubuntu and inspec goes all the way down when checking symlinks
  end

  # Verify the timezone file is correct
  describe file('/etc/timezone') do
    it { should exist }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0644 }
    its(:content) { should match /America\/New_York/ }
  end

  # Verify the ntp service is running and enabled
  describe service('ntp') do
    it { should be_enabled }
    it { should be_running }
  end

elsif os[:name] == 'centos'

  # Verify the NTP config exists
  describe file('/etc/ntp.conf') do
    it { should exist }
    it { should be_owned_by 'root' }
    it { should be_grouped_into 'root' }
    it { should be_mode 0644 }
  end

  # Verify the ntp package is installed
  describe package('ntp') do
    it { should be_installed }
  end

  # Verify the localtime is set correctly
  describe file('/etc/localtime') do
    it { should be_symlink }
    it { should be_linked_to '/usr/share/zoneinfo/America/New_York' }
  end

  # Verify the ntp service is running and enabled
  describe service('ntpd') do
    it { should be_enabled }
    it { should be_running }
  end
else
  # Do nothing
end


metadata.rb
name 'r1337-linux-base'
maintainer 'Matthew Ahrenstein'
maintainer_email 'matthew@route1337.com'
license 'MIT'
description 'Confgure basic settings, packages and services on Linux servers'
long_description 'This Chef cookbook will configure some basic system settings, add/remove some common packages, and deploy standard configurations for some common services'
version '0.1.7'
chef_version '>= 13.6.4' if respond_to?(:chef_version)
issues_url 'https://github.com/route1337/chef-cookbook-linuxbase/issues'
source_url 'https://github.com/route1337/chef-cookbook-linuxbase'
depends          'apt'
depends          'yum-epel'

%w{ ubuntu centos }.each do |os|
  supports os
end

recipe 'r1337-linux-base::default', 'Call this recipe to run the various tweaks from the cookbook'
recipe 'r1337-linux-base::prereqs', 'Called by default. Run all the prerequisites to system management'
recipe 'r1337-linux-base::bash', 'Called by default. Configures the custom bash profile'
recipe 'r1337-linux-base::thefuck', 'Called by default. Installs and configures thefuck'
recipe 'r1337-linux-base::ntp', 'Called by default. Installs and configures ntp for the New York timezone'


recipes/bash.rb
#
# Cookbook Name:: r1337-linux-base
# Recipe:: bash
#
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

# Remove root's bash
case node['platform']
  when 'ubuntu'
    file '/root/.bashrc' do
      action :delete
    end
  when 'centos'
    # Do nothing
  else
    # Do nothing
end

# Set the custom bashrc path
case node['platform']
  when 'ubuntu'
    bashpath =  '/etc/bash.bashrc'
  when 'centos'
    bashpath = '/etc/bashrc'
  else
    bashpath = '/etc/bash.bashrc' # Just to avoid throwing errors
end

# Deploy the custom bashrc and required git.bash
case node['platform']
  when 'ubuntu', 'centos'
    template "#{bashpath}" do
      source 'bash.bashrc.erb'
      owner 'root'
      group 'root'
      mode 0644
    end

    template '/opt/git.bash' do
      source 'git.bash.erb'
      owner 'root'
      group 'root'
      mode 0644
    end
  else
    # Do nothing
end


recipes/prereqs.rb
#
# Cookbook Name:: r1337-linux-base
# Recipe:: prereqs
#
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

# Run apt cookbook to refresh apt cache on Ubuntu-based systems
case node['platform']
  when 'ubuntu'
    include_recipe 'apt'
  else
    # Do nothing
end

# Make sure yum EPEL repo is added on CentOS-based systems
case node['platform']
  when 'centos'

    # Make sure the broken Fedora Rawhide repo is removed
    file '/etc/yum.repos.d/fedora-rawhide.repo' do
      action :delete
    end

    # Make sure the EPEL repo is added
    include_recipe 'yum-epel'
  else
    # Do nothing
end

# Purge unneeded default packages from the system based on OS
case node['platform']
  when 'ubuntu'
    # This creates an array of strings and performs the Chef package functionality of each array entry via a do loop
    %w{landscape-client-ui landscape-client-ui-install landscape-client landscape-common update-motd}.each do |pkg|
      package pkg do
        action :purge
      end
    end
  when 'centos'
    # No unneeded packages in CentOS right now
  else
    # Do nothing
end

# Install any needed packages we consider default
case node['platform']
  when 'ubuntu'
    %w{bmon htop vim curl gnupg2 atop git molly-guard python-pip python-dev python3-pip unzip psmisc}.each do |pkg|
      package pkg do
        action :install
      end
    end
  when 'centos'
    %w{vim-enhanced curl htop gnupg2 git lsof telnet net-tools bmon atop wget python2-pip python34-pip python34-devel unzip gcc psmisc bind-utils}.each do |pkg|
      package pkg do
        action :install
      end
    end
  else
    # Do nothing
end

# Make sure the tools directory exists
case node['platform']
  when 'ubuntu', 'centos'
    directory '/tools' do
      owner 'root'
      group 'root'
      mode 0700
      action :create
    end
  else
    # Do nothing
end

# Install any needed python 2.7 packages we consider default
case node['platform']
  when 'ubuntu'
    # Make sure python 2.7 has the latest pip
    execute 'Ensure latest pip2' do
      command '/usr/bin/python -m pip install --upgrade pip'
      not_if { File.exist?('/usr/lib/python2.7/dist-packages/pip') }
      action :run
    end
    # Specify the install manually since Chef does not have the best python management
    %w{urllib3 pyasn1}.each do |pypkg|
      execute "install #{pypkg}" do
        command "/usr/bin/python -m pip install --ignore-installed --upgrade #{pypkg}" # Let's just manually specify our python binary to be sure it's 2.7
        not_if { File.exist?("/usr/lib/python2.7/dist-packages/#{pypkg}") }
        action :run
      end
    end
  when 'centos'
    # Make sure python 2.7 has the latest pip
    execute 'Ensure latest pip2' do
      command '/usr/bin/python -m pip install --upgrade pip'
      not_if { File.exist?('/usr/lib/python2.7/site-packages/pip') }
      action :run
    end
    # Specify the install manually since Chef does not have the best python management
    %w{urllib3 pyasn1}.each do |pypkg|
      execute "install #{pypkg}" do
        command "/usr/bin/python -m pip install --ignore-installed --upgrade #{pypkg}" # Let's just manually specify our python binary to be sure it's 2.7
        not_if { File.exist?("/usr/lib/python2.7/site-packages/#{pypkg}") }
        action :run
      end
    end
  else
    # Do nothing
end


recipes/thefuck.rb
#
# Cookbook Name:: r1337-linux-base
# Recipe:: thefuck
#
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

# Install thefuck
case node['platform']
  when 'ubuntu'
    # Make sure python 3 has the latest pip
    execute 'Ensure latest pip3' do
      command '/usr/bin/python3 -m pip install --upgrade pip'
      not_if { File.exist?('/usr/lib/python3/dist-packages/pip') }
      action :run
    end

    # The below case block deals with Ubuntu 14.04 not having up to date python3 repos due to Canonical not keeping things updated
    case node['platform_version']
      when '14.04'
        # Specify thefuck to be installed locked to version 3.00 due to Canonical's poor repo management
        execute 'install thefuck' do
          command 'pip3 install thefuck==3.00' # Let's just manually specify our python binary to be sure it's 3
          not_if { File.exist?('/usr/local/bin/thefuck') }
          action :run
        end
      else
        # Specify the install manually since Chef does not have the best python management
        execute 'install thefuck' do
          command '/usr/bin/python3 -m pip install --ignore-installed --upgrade thefuck' # Let's just manually specify our python binary to be sure it's 3
          not_if { File.exist?('/usr/local/bin/thefuck') }
          action :run
        end
    end

    # Specify the install manually since Chef does not have the best python management
    execute 'install thefuck' do
      command 'pip3 install thefuck==3.00' # Let's just manually specify our python binary to be sure it's 3
      not_if { File.exist?('/usr/local/bin/thefuck') }
      action :run
    end
  when 'centos'
    # Make sure python 3 has the latest pip
    execute 'Ensure latest pip3' do
      command '/usr/bin/python3.4 -m pip install --upgrade pip'
      not_if { File.exist?('/usr/lib/python3.4/site-packages/pip') }
      action :run
    end
    # Specify the install manually since Chef does not have the best python management
    execute 'install thefuck' do
      command 'pip3 install --ignore-installed --upgrade thefuck' # Let's just manually specify our python binary to be sure it's 3
      not_if { File.exist?('/usr/bin/thefuck') }
      action :run
    end
  else
    # Do nothing
end

# Check out custom thefuck rules provided by Ahrenstein
case node['platform']
  when 'ubuntu', 'centos'
    # Create the directory structure for the repo
    %w{/root/.config /root/.config/thefuck /root/.config/thefuck/rules}.each do |confdir|
      directory confdir do
        owner 'root'
        group 'root'
        mode 0750
        action :create
      end
    end

    # Checkout the repi
    git 'thefuck rules' do
      destination '/root/.config/thefuck/rules'
      repository 'https://github.com/ahrenstein/thefuck-rules.git'
      revision 'master'
      action :sync
    end

    # Make sure the repo is not tampered with by resetting it
    execute 'git reset thefuck rules' do
      command 'git reset --hard HEAD && git clean -fd'
      cwd '/root/.config/thefuck/rules'
    end
  else
    # Do nothing
end


recipes/ntp.rb
#
# Cookbook Name:: r1337-linux-base
# Recipe:: ntp
#
# Copyright 2018, Route 1337, LLC, All Rights Reserved.
#
# Maintainers:
# - Matthew Ahrenstein: matthew@route1337.com
#
# See LICENSE
#

# Set the correct time zone for localtime
link 'localtime' do
  target_file '/etc/localtime'
  to '/usr/share/zoneinfo/America/New_York'
  link_type :symbolic
  action :create
end

# Ubuntu also needs the timezone file configured
case node['platform']
  when 'ubuntu'
    # Only Ubuntu uses the timezone file as well
    file 'timezone' do
      path '/etc/timezone'
      content 'America/New_York'
      force_unlink true
      owner 'root'
      group 'root'
      mode 0644
    end
  else
    # Do nothing
end

package 'ntp' do
  action :install
end

# Set variables for making sure the ntp service is started and enabled
case node['platform']
  when 'ubuntu'
    ntp_service_name = 'ntp'
  when 'centos'
    ntp_service_name = 'ntpd'
  else
    ntp_service_name = 'ntp' #We'll just assume everyone else uses ntp
end

# Make sure the ntp service is running and enabled
service ntp_service_name do
  action [ :enable, :start]
  subscribes :restart, 'file[/etc/ntp.conf', :immediately #Restart ntp service if Chef edits ntp.conf
end

# Deploy the ntp.conf
template '/etc/ntp.conf' do
  source 'ntp.conf.erb'
  owner 'root'
  group 'root'
  mode 0644
end
